stdlib/test/python/test_python_object.mojo

```diff
@@ -485,6 +485,93 @@ fn test_getitem_raises() raises:
         _ = with_2d[2]
 
 
+fn test_py_slice() raises:
+    var a = PythonObject([1, 2, 3, 4, 5])
+    assert_equal("[2, 3]", str(a[1:3]))
+    assert_equal("[1, 2, 3, 4, 5]", str(a[:]))
+    assert_equal("[1, 2, 3]", str(a[:3]))
+    assert_equal("[3, 4, 5]", str(a[2:]))
+    assert_equal("[1, 3, 5]", str(a[::2]))
+    assert_equal("[2, 4]", str(a[1::2]))
+    assert_equal("[4, 5]", str(a[-2:]))
+    assert_equal("[1, 2, 3]", str(a[:-2]))
+    assert_equal("[5, 4, 3, 2, 1]", str(a[::-1]))
+    assert_equal("[1, 2, 3, 4, 5]", str(a[-10:10]))  # out of bounds
+    assert_equal("[1, 2, 3, 4, 5]", str(a[::]))
+    assert_equal("[1, 2, 3, 4, 5]", str(a[:100]))
+    assert_equal("[]", str(a[5:]))
+    assert_equal("[5, 4, 3, 2]", str(a[:-5:-1]))
+
+    var b = Python.evaluate("[i for i in range(1000)]")
+    assert_equal("[0, 250, 500, 750]", str(b[::250]))
+    with assert_raises(contains="slice step cannot be zero"):
+        _ = b[::0]
+    # Negative cases such as `b[1.3:10]` or `b["1":10]` are handled by parser
+    # which would normally throw a TypeError in Python
+
+    var s = PythonObject("Hello, World!")
+    assert_equal("Hello", str(s[:5]))
+    assert_equal("World!", str(s[7:]))
+    assert_equal("!dlroW ,olleH", str(s[::-1]))
+    assert_equal("Hello, World!", str(s[:]))
+    assert_equal("Hlo ol!", str(s[::2]))
+    assert_equal("Hlo ol!", str(s[None:None:2]))
+
+    var t = PythonObject((1, 2, 3, 4, 5))
+    assert_equal("(2, 3, 4)", str(t[1:4]))
+    assert_equal("(4, 3, 2)", str(t[3:0:-1]))
+
+    var empty = PythonObject([])
+    assert_equal("[]", str(empty[:]))
+    assert_equal("[]", str(empty[1:2:3]))
+
+    var d = Python.dict()
+    d["a"] = 1
+    d["b"] = 2
+    with assert_raises(contains="slice(1, 3, 1)"):
+        _ = d[1:3]
+
+    var custom = Python.evaluate(
+        "type('CustomSliceable', (), {'__getitem__': lambda self, key: key})()"
+    )
+    assert_equal("slice(1, 3, 1)", str(custom[1:3]))
+
+    var i = PythonObject(1)
+    with assert_raises(contains="'int' object is not subscriptable"):
+        _ = i[0:1]
+
+    with_2d = Python.evaluate(
+        """type('With2D', (), {
+            '__init__': lambda self: setattr(self, 'data', [[1, 2, 3], [4, 5, 6], [7, 8, 9]]),
+            '__getitem__': lambda self, key: (
+                [row[key[1]] for row in self.data[key[0]]] if isinstance(key, tuple) and all(isinstance(k, slice) for k in key)
+                else (self.data[key[0]][key[1]] if isinstance(key, tuple)
+                else self.data[key])
+            )
+        })()"""
+    )
```

Comment:
Also, would integrating something like [`stdlib/test/python/my_module.py`](https://github.com/modularml/mojo/blob/nightly/stdlib/test/python/my_module.py) be preferred over this..?  @JoeLoser

Reply:
I think refactoring these sort of things to a separate test file (class with methods like you have here) may improve readability over raw strings.  Feel free to defer to a follow-up though :)